core/chains/evm/txm/storage/inmemory_store.go

```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	var unconfirmedTransactionIDs []uint64
+	for _, tx := range m.ConfirmedTransactions {
+		if tx.Nonce == nil {
+			return nil, nil, fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+		}
+		if *tx.Nonce >= latestNonce {
+			tx.State = types.TxUnconfirmed
+			tx.LastBroadcastAt = time.Time{} // Mark reorged transaction as if it wasn't broadcasted before
+			unconfirmedTransactionIDs = append(unconfirmedTransactionIDs, tx.ID)
+			m.UnconfirmedTransactions[*tx.Nonce] = tx
+			delete(m.ConfirmedTransactions, *tx.Nonce)
+		}
+	}
+
+	if len(m.ConfirmedTransactions) >= maxQueuedTransactions {
+		prunedTxIDs := m.pruneConfirmedTransactions()
+		m.lggr.Debugf("Confirmed transactions map for address: %v reached max limit of: %d. Pruned 1/3 of the oldest confirmed transactions. TxIDs: %v",
+			m.address, maxQueuedTransactions, prunedTxIDs)
+	}
+	sort.Slice(confirmedTransactions, func(i, j int) bool { return confirmedTransactions[i].ID < confirmedTransactions[j].ID })
+	sort.Slice(unconfirmedTransactionIDs, func(i, j int) bool { return unconfirmedTransactionIDs[i] < unconfirmedTransactionIDs[j] })
+	return confirmedTransactions, unconfirmedTransactionIDs, nil
+}
+
+func (m *InMemoryStore) MarkUnconfirmedTransactionPurgeable(nonce uint64) error {
+	m.Lock()
+	defer m.Unlock()
+
+	tx, exists := m.UnconfirmedTransactions[nonce]
+	if !exists {
+		return fmt.Errorf("unconfirmed tx with nonce: %d was not found", nonce)
+	}
+
+	tx.IsPurgeable = true
+
+	return nil
+}
+
+func (m *InMemoryStore) UpdateTransactionBroadcast(txID uint64, txNonce uint64, attemptHash common.Hash) error {
+	m.Lock()
+	defer m.Unlock()
+
+	unconfirmedTx, exists := m.UnconfirmedTransactions[txNonce]
+	if !exists {
+		return fmt.Errorf("unconfirmed tx was not found for nonce: %d - txID: %v", txNonce, txID)
+	}
+
+	// Set the same time for both the tx and its attempt
+	now := time.Now()
+	unconfirmedTx.LastBroadcastAt = now
+	if unconfirmedTx.InitialBroadcastAt.IsZero() {
+		unconfirmedTx.InitialBroadcastAt = now
+	}
+	a, err := unconfirmedTx.FindAttemptByHash(attemptHash)
+	if err != nil {
+		return err
+	}
+	a.BroadcastAt = now
+
+	return nil
+}
+
+func (m *InMemoryStore) UpdateUnstartedTransactionWithNonce(nonce uint64) (*types.Transaction, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	if len(m.UnstartedTransactions) == 0 {
+		m.lggr.Debugf("Unstarted transactions queue is empty for address: %v", m.address)
+		return nil, nil
+	}
+
+	if _, exists := m.UnconfirmedTransactions[nonce]; exists {
+		return nil, fmt.Errorf("an unconfirmed tx with the same nonce already exists: %v", m.UnconfirmedTransactions[nonce])
+	}
+
+	tx := m.UnstartedTransactions[0]
+	tx.Nonce = &nonce
+	tx.State = types.TxUnconfirmed
+
+	m.UnstartedTransactions = m.UnstartedTransactions[1:]
+	m.UnconfirmedTransactions[nonce] = tx
+
+	return tx.DeepCopy(), nil
+}
+
+// Shouldn't call lock because it's being called by a method that already has the lock
+func (m *InMemoryStore) pruneConfirmedTransactions() []uint64 {
```

Comment:
Could we maybe use a different naming convention to identify methods like this where the expectation is the caller has already acquired the lock? Think we had discussed something like this when Jim was working on the other in-memory work. Maybe we can lead with an `_`
```suggestion
func (m *InMemoryStore) _pruneConfirmedTransactions() []uint64 {
```

Reply:
I'm not a big fan of `_` since it comes from Python. I believe someone else made the same comment.

Reply:
We can leave it then. The comment is probably enough